plugins/woocommerce/src/Internal/Utilities/BatchProcessor.php

```diff
@@ -0,0 +1,66 @@
+<?php
+/**
+ * Implements abstract class for action scheduler based batch processing. Updates using this class must be tolerant of same record getting processed more than once.
+ */
+
+namespace Automattic\WooCommerce\Internal\Utilities;
+
+/**
+ * Class BatchProcessor
+ *
+ * @package Automattic\WooCommerce\DataBase
+ */
+abstract class BatchProcessor {
+
+	/**
+	 * A user friendly name for this process.
+	 *
+	 * @return string Name of the process.
+	 */
+	abstract protected function get_name() : string;
+
+	/**
+	 * A user friendly description for this process.
+	 *
+	 * @return string Description.
+	 */
+	abstract protected function get_description() : string;
+
+	/**
+	 * Process data for current batch.
+	 *
+	 * @param array $batch Batch details.
+	 */
+	abstract public function process_for_batch( array $batch );
+
+	/**
+	 * Get total number of pending records that require processing.
+	 *
+	 * @return int Number of pending records.
+	 */
+	abstract public function get_total_pending_count() : int;
+
+	/**
+	 * Returns the batch with records that needs to be processed for a given size.
+	 *
+	 * @param int   $size Size of the batch.
+	 * @param mixed $last_processed Identifier of record that was last processed.
+	 *
+	 * @return array Batch of records.
+	 */
+	abstract public function get_batch_data( int $size, $last_processed ) : array;
```

Comment:
About the method name: I still thing that `get_next_batch_to_process` better conveys the purpose of the method, as `get_batch_data` sounds too generic (batch data? Which batch data?)

About the `$last_processed` argument: I get the intent, you want to shield the processor class from having to store any kind of intermediate state. However I think there's a better way to achieve this: a `&$state` parameter passed by reference. Pass it as `null` the first time, let the processor set/alter it as it wishes, and just store it treating it as an opaque value. Bonus points if you add the same parameter to `process_for_batch` and to `get_total_pending_count` too.

Reply:
I have changed the method name in https://github.com/woocommerce/woocommerce/pull/33233/files#diff-1e070cc170d3b65cd6ba61b60cbd0697c8431fda674b4720926c8fbc1afecce7R297. I agree about the state param, but then we would have to manage it's safe serialization and vice-versa, so let's add it when we need it.

Reply:
I agree and lean more towards not having a state to pass at all. I have removed the `last_processed` argument in https://github.com/woocommerce/woocommerce/pull/33233/commits/311f014dc39a5cfe7be85e3eef6c332cddae4618